subsys/usb/class/mass_storage.c

```diff
@@ -290,74 +292,98 @@ static void testUnitReady(void)
 	sendCSW();
 }
 
-static bool requestSense(void)
+static struct fixed_format_sense_data sense_data = {
+	.code = SDRC_CURRENT_ERRORS,
+	.sense_key = SK_ILLEGAL_REQUEST,
+	.as_length = sizeof(struct additional_sense_data),
+	.asd = {
+		.asc_ascq = {0x30, 0x01},
+	},
+};
+
+static void update_sense_data_ascq(u16_t ascq)
+{
+	sys_put_be16(ascq, sense_data.asd.asc_ascq);
+}
+
+static bool req_sense_cmd(void)
 {
-	u8_t request_sense[] = {
-		0x70,
-		0x00,
-		0x05,   /* Sense Key: illegal request */
-		0x00,
-		0x00,
-		0x00,
-		0x00,
-		0x0A,
-		0x00,
-		0x00,
-		0x00,
-		0x00,
-		0x30,
-		0x01,
-		0x00,
-		0x00,
-		0x00,
-		0x00,
-	};
-
-	return write(request_sense, sizeof(request_sense));
+	struct cdb6 *cmd = (struct cdb6 *)&cbw.CB[0];
+
+	return write((u8_t *)&sense_data,
+		     MIN(cdb6_get_length(cmd), sizeof(sense_data)));
 }
 
-static bool inquiryRequest(void)
+static struct dabc_inquiry_data inq_data = {
+	.type = DIRECT_ACCESS_BLOCK_DEVICE,
+	.qualifier = 0,
+	.rmb = 1,	/* Medium is not removable */
+	.version = 0,
+	.rdf = 2,
+	.length = sizeof(struct dabc_inquiry_data) - 5,
+	.sccs = 0,
+	.t10_vid = {'Z', 'E', 'P', 'H', 'Y', 'R', ' ', ' '},
+	.product_id = {'Z', 'E', 'P', 'H', 'Y', 'R', ' ', 'U', 'S', 'B', ' ',
+		       'D', 'I', 'S', 'K', ' '},
+	.product_rev = {'0', '.', '0', '1'},
+};
+
+static bool inquiry_cmd(void)
 {
-	u8_t inquiry[] = { 0x00, 0x80, 0x00, 0x01,
-	36 - 4, 0x80, 0x00, 0x00,
-	'Z', 'E', 'P', 'H', 'Y', 'R', ' ', ' ',
-	'Z', 'E', 'P', 'H', 'Y', 'R', ' ', 'U', 'S', 'B', ' ',
-	'D', 'I', 'S', 'K', ' ',
-	'0', '.', '0', '1',
-	};
-
-	return write(inquiry, sizeof(inquiry));
+	struct cdb_inquiry *cmd = (struct cdb_inquiry *)&cbw.CB[0];
+
+	if (cmd->evpd) {
+		update_sense_data_ascq(ASCQ_INVALID_FIELD_IN_CDB);
+		csw.Status = CSW_FAILED;
+		sendCSW();
+		return false;
+	}
+
+	return write((u8_t *)&inq_data,
+		     MIN(sys_get_be16(&cmd->length[0]), sizeof(inq_data)));
```

Comment:
Two points here:
1) Shouldn't this be inquiry command 4th Byte? (counted from 0)
``` diff
-    MIN(sys_get_be16(&cmd->length[0]), sizeof(inq_data)));
+    MIN(sys_get_be16(&cmd->length[1]), sizeof(inq_data)));
```
2) I have concerns how this will behave if ``` &cmd->length[0] will actually be 0```.
Shouldn't we in this case cass sendCSW() directly? It seems to me there will not be data stage if inquiry command has length == 0. Just wondering what do You think.